org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testStopOutlierDetectionWithEnoughDocumentsToScroll fails on master

org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testStopOutlierDetectionWithEnoughDocumentsToScroll failed on master:
```
13:40:39 org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT > testStopOutlierDetectionWithEnoughDocumentsToScroll FAILED
13:40:39     java.lang.AssertionError: 
13:40:39     Expected: <0L>
13:40:39          but: was <1L>
13:40:39         at __randomizedtesting.SeedInfo.seed([CE9E38CE5E762526:5E6FCDEAD92BA21C]:0)
13:40:39         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
13:40:39         at org.junit.Assert.assertThat(Assert.java:956)
13:40:39         at org.junit.Assert.assertThat(Assert.java:923)
13:40:39         at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.cleanUpAnalytics(MlNativeDataFrameAnalyticsIntegTestCase.java:52)
13:40:39         at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.cleanUpResources(MlNativeDataFrameAnalyticsIntegTestCase.java:45)
13:40:39         at org.elasticsearch.xpack.ml.integration.MlNativeIntegTestCase.cleanUp(MlNativeIntegTestCase.java:127)
13:40:39         at org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.cleanup(RunDataFrameAnalyticsIT.java:46)
```

The assertion fails during cleanup:
```
                assertThat(searchStoredProgress(config.getId()).getHits().getTotalHits().value, equalTo(0L));
```

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:integTestRunner' --tests "org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testStopOutlierDetectionWithEnoughDocumentsToScroll" \
  -Dtests.seed=CE9E38CE5E762526 \
  -Dtests.security.manager=true \
  -Dtests.locale=en-CC \
  -Dtests.timezone=Asia/Choibalsan \
  -Dcompiler.java=12 \
  -Druntime.java=11
```
Does not reproduce locally.


Build | 20190913073350-B735813E
-- | --
Log | https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-darwin-compatibility/207/console
Build Scans | https://gradle-enterprise.elastic.co/s/z4parnu5jlhoyhttps://gradle-enterprise.elastic.co/s/ti53rgd66ecokhttps://gradle-enterprise.elastic.co/s/dqgzy2f5yitna

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java
@@ -7,6 +7,7 @@ package org.elasticsearch.xpack.ml.dataframe;
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ResourceNotFoundException;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.admin.indices.create.CreateIndexResponse;
 import org.elasticsearch.action.admin.indices.delete.DeleteIndexAction;
@@ -78,7 +79,13 @@ public class DataFrameAnalyticsManager {
                     case STARTED:
                         task.updatePersistentTaskState(reindexingState, ActionListener.wrap(
                             updatedTask -> reindexingStateListener.onResponse(config),
-                            reindexingStateListener::onFailure));
+                            error -> {
+                                if (error instanceof ResourceNotFoundException) {
+                                    // The task has been stopped
+                                } else {
+                                    reindexingStateListener.onFailure(error);
+                                }
+                            }));
                         break;
                     // The task has fully reindexed the documents and we should continue on with our analyses
                     case ANALYZING:
@@ -221,7 +228,13 @@ public class DataFrameAnalyticsManager {
                                 task.markAsCompleted();
                             }
                         }),
-                    error -> task.updateState(DataFrameAnalyticsState.FAILED, error.getMessage())
+                    error -> {
+                        if (error instanceof ResourceNotFoundException) {
+                            // Task has stopped
+                        } else {
+                            task.updateState(DataFrameAnalyticsState.FAILED, error.getMessage());
+                        }
+                    }
                 ));
             },
             error -> task.updateState(DataFrameAnalyticsState.FAILED, error.getMessage())

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java
@@ -96,7 +96,12 @@ public class DataFrameAnalyticsTask extends AllocatedPersistentTask implements S
 
     @Override
     public void markAsCompleted() {
-        persistProgress(() -> super.markAsCompleted());
+        // It is possible that the stop API has been called in the meantime and that
+        // may also cause this method to be called. We check whether we have already
+        // been marked completed to avoid doing it twice.
+        if (isCompleted() == false) {
+            persistProgress(() -> super.markAsCompleted());
+        }
     }
 
     @Override

